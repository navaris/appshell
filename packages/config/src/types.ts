import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';

export type Schema = JSONSchema4 | JSONSchema6 | JSONSchema7;

export type ConfigValidator = {
  validate: <T>(...config: T[]) => void;
};

export type ConfigMap = Record<string, string>;

/* appshell.config.yaml types */
export type AppshellConfigRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  metadata: TMetadata;
};

export type AppshellConfig<TMetadata = Record<string, unknown>> = {
  name?: string;
  remotes?: Record<string, AppshellConfigRemote<TMetadata>>;
  module: ModuleFederationPluginOptions;
};

/** Appshell manifest types */
export type AppshellRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  scope: string;
  module: string;
  shareScope?: string;
  metadata: TMetadata;
};

export type AppshellManifest<TMetadata = Record<string, unknown>> = {
  remotes: Record<string, AppshellRemote<TMetadata>>;
  modules: Record<string, ModuleFederationPluginOptions>;
};

/**
 * Advanced configuration for modules that should be exposed by this container.
 */
export interface ExposesConfig {
  /**
   * Request to a module that should be exposed by this container.
   */
  import: string | string[];

  /**
   * Custom chunk name for the exposed module.
   */
  name?: string;
}

export interface ExposesObject {
  [index: string]: string | ExposesConfig | string[];
}

/**
 * Set explicit comments for `commonjs`, `commonjs2`, `amd`, and `root`.
 */
export interface LibraryCustomUmdCommentObject {
  /**
   * Set comment for `amd` section in UMD.
   */
  amd?: string;

  /**
   * Set comment for `commonjs` (exports) section in UMD.
   */
  commonjs?: string;

  /**
   * Set comment for `commonjs2` (module.exports) section in UMD.
   */
  commonjs2?: string;

  /**
   * Set comment for `root` (global variable) section in UMD.
   */
  root?: string;
}

/**
 * Description object for all UMD variants of the library name.
 */
export interface LibraryCustomUmdObject {
  /**
   * Name of the exposed AMD library in the UMD.
   */
  amd?: string;

  /**
   * Name of the exposed commonjs export in the UMD.
   */
  commonjs?: string;

  /**
   * Name of the property exposed globally by a UMD library.
   */
  root?: string | string[];
}

/**
 * Options for library.
 */
export interface LibraryOptions {
  /**
   * Add a comment in the UMD wrapper.
   */
  auxiliaryComment?: string | LibraryCustomUmdCommentObject;

  /**
   * Specify which export should be exposed as library.
   */
  export?: string | string[];

  /**
   * The name of the library (some types allow unnamed libraries too).
   */
  name?: string | string[] | LibraryCustomUmdObject;

  /**
   * Type of library (types included by default are 'var', 'module', 'assign', 'assign-properties', 'this', 'window', 'self', 'global', 'commonjs', 'commonjs2', 'commonjs-module', 'commonjs-static', 'amd', 'amd-require', 'umd', 'umd2', 'jsonp', 'system', but others might be added by plugins).
   */
  type: string;

  /**
   * If `output.libraryTarget` is set to umd and `output.library` is set, setting this to true will name the AMD module.
   */
  umdNamedDefine?: boolean;
}

export interface SharedConfig {
  /**
   * Include the provided and fallback module directly instead behind an async request. This allows to use this shared module in initial load too. All possible shared modules need to be eager too.
   */
  eager?: boolean;

  /**
   * Provided module that should be provided to share scope. Also acts as fallback module if no shared module is found in share scope or version isn't valid. Defaults to the property name.
   */
  import?: string | false;

  /**
   * Package name to determine required version from description file. This is only needed when package name can't be automatically determined from request.
   */
  packageName?: string;

  /**
   * Version requirement from module in share scope.
   */
  requiredVersion?: string | false;

  /**
   * Module is looked up under this key from the share scope.
   */
  shareKey?: string;

  /**
   * Share scope name.
   */
  shareScope?: string;

  /**
   * Allow only a single version of the shared module in share scope (disabled by default).
   */
  singleton?: boolean;

  /**
   * Do not accept shared module if version is not valid (defaults to yes, if local fallback module is available and shared module is not a singleton, otherwise no, has no effect if there is no required version specified).
   */
  strictVersion?: boolean;

  /**
   * Version of the provided module. Will replace lower matching versions, but not higher.
   */
  version?: string | false;
}

export interface SharedObject {
  [index: string]: string | SharedConfig;
}

/**
 * Advanced configuration for container locations from which modules should be resolved and loaded at runtime.
 */
export interface RemotesConfig {
  /**
   * Container locations from which modules should be resolved and loaded at runtime.
   */
  external: string | string[];

  /**
   * The name of the share scope shared with this remote.
   */
  shareScope?: string;
}

/**
 * Container locations from which modules should be resolved and loaded at runtime. Property names are used as request scopes.
 */
export interface RemotesObject {
  [index: string]: string | RemotesConfig | string[];
}

export interface ModuleFederationPluginOptions {
  /**
   * Modules that should be exposed by this container. When provided, property name is used as public name, otherwise public name is automatically inferred from request.
   */
  exposes?: (string | ExposesObject)[] | ExposesObject;

  /**
   * The filename of the container as relative path inside the `output.path` directory.
   */
  filename?: string;

  /**
   * Options for library.
   */
  library?: LibraryOptions;

  /**
   * The name of the container.
   */
  name?: string;

  /**
   * The external type of the remote containers.
   */
  remoteType?:
    | 'import'
    | 'var'
    | 'module'
    | 'assign'
    | 'this'
    | 'window'
    | 'self'
    | 'global'
    | 'commonjs'
    | 'commonjs2'
    | 'commonjs-module'
    | 'commonjs-static'
    | 'amd'
    | 'amd-require'
    | 'umd'
    | 'umd2'
    | 'jsonp'
    | 'system'
    | 'promise'
    | 'script'
    | 'node-commonjs';

  /**
   * Container locations and request scopes from which modules should be resolved and loaded at runtime. When provided, property name is used as request scope, otherwise request scope is automatically inferred from container location.
   */
  remotes?: (string | RemotesObject)[] | RemotesObject;

  /**
   * The name of the runtime chunk. If set a runtime chunk with this name is created or an existing entrypoint is used as runtime.
   */
  runtime?: string | false;

  /**
   * Share scope name used for all shared modules (defaults to 'default').
   */
  shareScope?: string;

  /**
   * Modules that should be shared in the share scope. When provided, property names are used to match requested modules in this compilation.
   */
  shared?: (string | SharedObject)[] | SharedObject;
}
