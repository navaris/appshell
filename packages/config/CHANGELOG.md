# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.5.2](https://github.com/navaris/appshell/compare/@appshell/config@0.5.1...@appshell/config@0.5.2) (2023-08-29)


### Bug Fixes

* appshell loader override check ([aedda4a](https://github.com/navaris/appshell/commit/aedda4abf81560297dce03df09fba68e8785242f))





## [0.5.1](https://github.com/navaris/appshell/compare/@appshell/config@0.5.0...@appshell/config@0.5.1) (2023-08-29)


### Bug Fixes

* override support for appshell template schema ([f2b10d4](https://github.com/navaris/appshell/commit/f2b10d4e784d2d39b4602390b73b7db89bd88c34))





# [0.5.0](https://github.com/navaris/appshell/compare/@appshell/config@0.5.0-alpha.1...@appshell/config@0.5.0) (2023-08-28)

**Note:** Version bump only for package @appshell/config





# [0.5.0-alpha.1](https://github.com/navaris/appshell/compare/@appshell/config@0.5.0-alpha.0...@appshell/config@0.5.0-alpha.1) (2023-08-28)


### Features

* environment variable override for federated components ([#27](https://github.com/navaris/appshell/issues/27)) ([d32de0b](https://github.com/navaris/appshell/commit/d32de0b0d1cbb1792715e1b363c80ed4600df155))





# [0.5.0-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.4.0...@appshell/config@0.5.0-alpha.0) (2023-08-23)


### Features

* consolidate runtime artifacts ([#26](https://github.com/navaris/appshell/issues/26)) ([a29479a](https://github.com/navaris/appshell/commit/a29479a49f0c5ec1273c9f8e4c7384096f2d4ba0))





# [0.4.0](https://github.com/navaris/appshell/compare/@appshell/config@0.4.0-alpha.0...@appshell/config@0.4.0) (2023-08-09)

**Note:** Version bump only for package @appshell/config





# [0.4.0-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.3.2...@appshell/config@0.4.0-alpha.0) (2023-08-09)


### Features

* deregister cli command ([#24](https://github.com/navaris/appshell/issues/24)) ([746e827](https://github.com/navaris/appshell/commit/746e8273b366543a606cecadf18554bfc094143e))





## [0.3.2](https://github.com/navaris/appshell/compare/@appshell/config@0.3.2-alpha.0...@appshell/config@0.3.2) (2023-07-18)

**Note:** Version bump only for package @appshell/config





## [0.3.2-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.3.1...@appshell/config@0.3.2-alpha.0) (2023-07-18)


### Bug Fixes

* schema validation only when registering manifests ([c0136fb](https://github.com/navaris/appshell/commit/c0136fb1cc9163e575c233e8330593a2a6c5a670))





## [0.3.1](https://github.com/navaris/appshell/compare/@appshell/config@0.3.1-alpha.0...@appshell/config@0.3.1) (2023-07-17)

**Note:** Version bump only for package @appshell/config





## [0.3.1-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0...@appshell/config@0.3.1-alpha.0) (2023-07-17)


### Bug Fixes

* enable insecure flag for cli generate index, metadata and start commands ([03f5e7f](https://github.com/navaris/appshell/commit/03f5e7fbd2ad3746ce648152c4410025958274a9))





# [0.3.0](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.7...@appshell/config@0.3.0) (2023-07-11)


### Bug Fixes

* reenable unit tests ([c00f581](https://github.com/navaris/appshell/commit/c00f581b2d71778378729f3bc42a4f5a1c0afe04))





# [0.3.0-alpha.7](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.6...@appshell/config@0.3.0-alpha.7) (2023-07-10)


### Bug Fixes

* replace config intermediate file with template ([9b79e65](https://github.com/navaris/appshell/commit/9b79e65b355686a0cca273c89c7164bb031e8437))





# [0.3.0-alpha.6](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.5...@appshell/config@0.3.0-alpha.6) (2023-07-10)


### Bug Fixes

* update registy url for metadata ([20a9229](https://github.com/navaris/appshell/commit/20a92296289d3510ff9661654bac303c0ccd07b7))





# [0.3.0-alpha.5](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.4...@appshell/config@0.3.0-alpha.5) (2023-07-10)


### Bug Fixes

* ship developer image with development build, urls for adjunct registries ([333f931](https://github.com/navaris/appshell/commit/333f9314854a93ab3f8e57dfde806baf26012e6c))





# [0.3.0-alpha.4](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.3...@appshell/config@0.3.0-alpha.4) (2023-07-10)


### Bug Fixes

* registry projection, enable env overwrite, refactor docker workflow, consolidate env names ([33a28fe](https://github.com/navaris/appshell/commit/33a28fe76b58e05c5b6b6b33d4b402e52bb29e70))





# [0.3.0-alpha.3](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.2...@appshell/config@0.3.0-alpha.3) (2023-07-06)


### Bug Fixes

* support metadata flag on cli start ([567cf02](https://github.com/navaris/appshell/commit/567cf02a52150054b855197371681f426382b454))





# [0.3.0-alpha.2](https://github.com/navaris/appshell/compare/@appshell/config@0.3.0-alpha.1...@appshell/config@0.3.0-alpha.2) (2023-07-06)


### Features

* appshell start, cleanup dev workflow ([74fd2e5](https://github.com/navaris/appshell/commit/74fd2e5a5acd2415482268175c7f3f16cd7c93ec))





# [0.3.0-alpha.1](https://github.com/navaris/appshell/compare/@appshell/config@0.2.1...@appshell/config@0.3.0-alpha.1) (2023-07-05)


### Features

* foundation refactor wip ([339b930](https://github.com/navaris/appshell/commit/339b9306accaa9aba1712f3a0dee6c4cab7ed273))





# [0.3.0-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.2.1...@appshell/config@0.3.0-alpha.0) (2023-07-05)


### Bug Fixes

* package dependencies ([49f1ace](https://github.com/navaris/appshell/commit/49f1ace6fd63b35ea3c866315123054517687158))


### Features

* registry index ([8e6ee0a](https://github.com/navaris/appshell/commit/8e6ee0a6a377584efa2ee702168025f46108b8c5))





## [0.2.1](https://github.com/navaris/appshell/compare/@appshell/config@0.2.1-alpha.2...@appshell/config@0.2.1) (2023-06-29)

**Note:** Version bump only for package @appshell/config





## [0.2.1-alpha.2](https://github.com/navaris/appshell/compare/@appshell/config@0.2.0...@appshell/config@0.2.1-alpha.2) (2023-06-29)

**Note:** Version bump only for package @appshell/config






## [0.2.1-alpha.1](https://github.com/navaris/appshell/compare/@appshell/config@0.2.1-alpha.0...@appshell/config@0.2.1-alpha.1) (2023-06-29)


### Bug Fixes

* update npm package visibility ([1ef4119](https://github.com/navaris/appshell/commit/1ef411903dd038dfc781e8ce0700811e5460c903))





## [0.2.1-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.2.0-alpha.2...@appshell/config@0.2.1-alpha.0) (2023-06-29)

**Note:** Version bump only for package @appshell/config





# [0.2.0](https://github.com/navaris/appshell/compare/@appshell/config@0.2.0-alpha.2...@appshell/config@0.2.0) (2023-06-29)

**Note:** Version bump only for package @appshell/config





# [0.2.0-alpha.2](https://github.com/navaris/appshell/compare/@appshell/config@0.2.0-alpha.1...@appshell/config@0.2.0-alpha.2) (2023-06-28)


### Bug Fixes

* Fixed parsing of strings with more than one variable. ([#10](https://github.com/navaris/appshell/issues/10)) ([af21588](https://github.com/navaris/appshell/commit/af215886b858742ecf170a6e06a1d6bcba3fd39c))





# [0.2.0-alpha.1](https://github.com/navaris/appshell/compare/@appshell/config@0.1.7-alpha.0...@appshell/config@0.2.0-alpha.1) (2023-06-22)


### Features

* support manifest merging and config watch ([5431d10](https://github.com/navaris/appshell/commit/5431d100ec7f5106cab66d3009ce5f836e452715))





## [0.1.7-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.1.6...@appshell/config@0.1.7-alpha.0) (2023-06-15)

**Note:** Version bump only for package @appshell/config





## [0.1.6](https://github.com/navaris/appshell/compare/@appshell/config@0.1.5...@appshell/config@0.1.6) (2023-06-15)

**Note:** Version bump only for package @appshell/config

## [0.1.6-alpha.2](https://github.com/navaris/appshell/compare/@appshell/config@0.1.6-alpha.1...@appshell/config@0.1.6-alpha.2) (2023-06-15)

## [0.1.6-alpha.1](https://github.com/navaris/appshell/compare/@appshell/config@0.1.6-alpha.0...@appshell/config@0.1.6-alpha.1) (2023-03-19)

## [0.1.6-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.1.5-alpha.0...@appshell/config@0.1.6-alpha.0) (2023-03-19)

**Note:** Version bump only for package @appshell/config

## [0.1.5](https://github.com/navaris/appshell/compare/@appshell/config@0.1.5-alpha.0...@appshell/config@0.1.5) (2023-03-05)

**Note:** Version bump only for package @appshell/config

## [0.1.5-alpha.0](https://github.com/navaris/appshell/compare/@appshell/config@0.1.4...@appshell/config@0.1.5-alpha.0) (2023-03-05)

**Note:** Version bump only for package @appshell/config

### Bug Fixes

- add shareScope support to appshell remotes ([ceb55c8](https://github.com/navaris/appshell/commit/ceb55c8f6deed27ad11cff841609b1656269a921))

## [0.1.4](https://github.com/navaris/appshell/compare/@appshell/config@0.1.4-alpha.0...@appshell/config@0.1.4) (2023-03-04)

**Note:** Version bump only for package @appshell/config

## 0.1.4-alpha.0 (2023-03-04)

### Bug Fixes

- setup github actions build pipeline, setup project for release ([a54722f](https://github.com/navaris/appshell/commit/a54722f3df28098593ec1bce3cc2def377ff531a))
