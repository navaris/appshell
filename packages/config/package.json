{
  "name": "@appshell/config",
  "version": "0.2.1-alpha.1",
  "description": "Utility for producing a global appshell manifest for module federation micro-frontends",
  "main": "dist/main.js",
  "types": "dist/types/index.d.ts",
  "repository": "https://github.com/navaris/appshell.git",
  "author": "Robert Hamilton <rh@navaris.com>",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "scripts": {
    "build": "webpack --mode production && yarn build:types",
    "build:types": "tsc --emitDeclarationOnly",
    "clean": "rm -rf dist",
    "start": "webpack --watch --mode development && yarn build:types",
    "stats": "webpack --json > stats.json"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "webpack",
    "module federation",
    "micro-frontends",
    "appshell"
  ],
  "license": "MIT"
}
