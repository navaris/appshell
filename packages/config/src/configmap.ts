import { entries, set, values } from 'lodash';
import { AppshellConfig, ConfigMap } from './types';

const matchVariable = /(\${\w+})/;
const matchWord = /[^\w @]/g;

const findVariables = (val: string) =>
  val
    .split(matchVariable)
    .filter((w) => w.match(matchVariable))
    ?.map((w) => w.replace(matchWord, ''));

const findVariablePlaceholders = (
  obj: object | string | number | undefined,
  results: Record<string, string> = {},
): Record<string, string> =>
  values(obj).reduce((acc: Record<string, string>, val: object | string | number | undefined) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      return VARS.length > 0 ? VARS.reduce((a, v) => set<string>(a, v, process.env[v]), acc) : acc;
    }
    if (typeof val === 'object') {
      return findVariablePlaceholders(val, acc);
    }
    return acc;
  }, results);

const apply = <T extends object>(obj: T, configMap: ConfigMap): T => {
  entries(obj).forEach(([key, val]) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      if (VARS.length > 0) {
        VARS.forEach((v) => {
          const cur = obj[key];
          const value = configMap[v];
          if (!value) {
            // eslint-disable-next-line no-console
            console.log(`warning: value for ${v} is ${value}`);
          }
          set<string>(obj, key, cur.replace(`$\{${v}}`, configMap[v]));
        });
      }
    } else if (typeof val === 'object') {
      apply(val, configMap);
    }
  });

  return obj;
};

const create = (config: AppshellConfig) => findVariablePlaceholders(config);

export default {
  apply,
  create,
};
