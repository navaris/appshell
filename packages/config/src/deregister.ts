/* eslint-disable no-param-reassign */
import fs from 'fs';
import axios from './axios';
import { AppshellManifest, AppshellRegister } from './types';
import { isValidUrl } from './utils';
import * as validators from './validators';

const removeFromRegister = (doc: AppshellRegister, key: string) => {
  if (doc.index[key]) {
    delete doc.index[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`index entry not found for '${key}'`);
  }

  if (doc.metadata && doc.metadata[key]) {
    delete doc.metadata[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`metadata entry not found for '${key}'`);
  }

  return doc;
};

const removeFromManifest = (doc: AppshellManifest, key: string) => {
  const targetModule = key.split('/')[0];

  if (!doc.remotes[key]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for remotes[${key}]`);
  }

  if (!doc.modules[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for modules[${key}]`);
  }

  if (!doc.environment[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for environment[${key}]`);
  }

  delete doc.remotes[key];
  delete doc.modules[targetModule];
  delete doc.environment[targetModule];

  return doc;
};

const updateRegister = (registry: string, key: string) => {
  const file = `${registry}/appshell.register.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const register: AppshellRegister = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromRegister(register, key);

  validators.appshell_register.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

const updateManifest = (registry: string, key: string) => {
  const file = `${registry}/appshell.manifest.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const manifest: AppshellManifest = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromManifest(manifest, key);

  validators.merge_manifests.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

export default async (manifestKey: string, registry: string) => {
  // eslint-disable-next-line no-console
  console.log(`deregistering manifest from registry ${registry}`);

  if (isValidUrl(registry)) {
    await axios.delete(registry, { data: { manifestKey } });

    return;
  }

  if (!fs.existsSync(registry)) {
    // eslint-disable-next-line no-console
    console.warn(`registry location not found ${registry}`);
  }

  updateRegister(registry, manifestKey);
  updateManifest(registry, manifestKey);
};
