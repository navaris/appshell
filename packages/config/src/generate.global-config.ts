/* eslint-disable no-console */
import { AppshellGlobalConfig } from './types';
import { isValidUrl, merge } from './utils';
import loadJson from './utils/loadJson';
import { AppshellGlobalConfigValidator } from './validators';

type GenerateGlobalConfigOptions = {
  insecure: boolean;
};

export default async (
  registries: string[],
  options: GenerateGlobalConfigOptions = { insecure: false },
): Promise<AppshellGlobalConfig> => {
  const defaultGlobalConfig = { index: {}, metadata: {}, overrides: { environment: {} } };

  if (registries.length < 1) {
    console.log(`No registries found. skipping global configuration generation.`);
    return defaultGlobalConfig;
  }

  console.log(
    `generating global appshell configuration --registry=${JSON.stringify(registries, null, 2)}`,
  );

  try {
    const registers = await Promise.all(
      registries.map((reg) => {
        const registry = isValidUrl(reg) ? `${reg}/appshell.config.json` : reg;
        return loadJson<AppshellGlobalConfig>(registry, {
          insecure: options.insecure,
          target: /(.config.json)/i,
        });
      }),
    ).then((items) => items.flat());
    console.log(
      `Generating global configuration from ${registers.length} source${
        registers.length === 1 ? '' : 's'
      }`,
    );

    return merge(AppshellGlobalConfigValidator, ...registers) as AppshellGlobalConfig;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating global appshell configuration', err.message);
  }

  return defaultGlobalConfig;
};
