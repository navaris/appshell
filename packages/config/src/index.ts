/**
 * @appshell/config package API
 */
export { default as configmap } from './configmap';
export { default as deregister } from './deregister';
export { default as generateEnv } from './generate.env';
export { default as generateGlobalConfig } from './generate.global-config';
export { default as generateManifest } from './generate.manifest';
export { default as register } from './register';
export type {
  AppshellConfig,
  AppshellConfigRemote,
  AppshellIndex,
  AppshellManifest,
  AppshellRemote,
  AppshellTemplate,
  Metadata,
  Schema,
} from './types';
export * as utils from './utils';
export * as validators from './validators';
