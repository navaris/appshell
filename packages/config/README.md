<div align="center">
  <a href="https://github.com/navaris/appshell">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo-white_2x.png">
      <img alt="appshell" src="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo_2x.png">
    </picture>
  </a>
</div>

[![Appshell CI](https://github.com/navaris/appshell/actions/workflows/pipeline.yml/badge.svg)](https://github.com/navaris/appshell/actions/workflows/pipeline.yml)

# @appshell/config

Utitliy for generating a `global runtime manifest` for Webpack Module federation micro-frontends.

Working examples can be found [here](https://github.com/navaris/appshell/tree/main/examples).

## Getting Started

To begin, you'll need to install `@appshell/config`:

```console
npm install @appshell/config --save-dev
```

or

```console
yarn add -D @appshell/config
```

or

```console
pnpm add -D @appshell/config
```

The primary export from this package is the `generate` function. It is given a configs dir to process and produces a global runtime manifest.

```ts
import { generate } from '@appshell/config';

const manifest = generate<MyMetadata>(process.env.APPSHELL_CONFIGS_DIR);
```

![Sample APPSHELL_CONFIGS_DIR](https://github.com/navaris/appshell/blob/main/assets/docs/appshell_configs_dir.png 'APPSHELL_CONFIGS_DIR')

**Where does the content of APPSHELL_CONFIGS_DIR come from?**

> Each micro-frontend configured to use [@appshell/manifest-webpack-plugin](https://www.npmjs.com/package/@appshell/manifest-webpack-plugin) emits it's configuration to the configs directory at build time, which is subsequently processed with this utility to reflect the current runtime environment.

Sample content from APPSHELL_CONFIGS_DIR:

```json
{
  "remotes": {
    "CraModule/App": {
      "url": "${CRA_MFE_URL}",
      "metadata": {
        "route": "/cra",
        "displayName": "Example App",
        "displayGroup": "${CRA_MFE_DISPLAY_GROUP}",
        "order": 10,
        "icon": "ViewList"
      },
      "id": "3eb81a0c"
    }
  },
  "module": {
    "exposes": {
      "./App": "./src/App"
    },
    "filename": "remoteEntry.js",
    "name": "CraModule",
    "shared": {
      "react": {
        "singleton": true,
        "requiredVersion": "^18.2.0"
      },
      "react-dom": {
        "singleton": true,
        "requiredVersion": "^18.2.0"
      }
    }
  }
}
```

**How does my runtime environment get reflected in the global runtime manifest?**

> Note the variable expansion syntax `${CRA_MFE_URL}`. When `generate` is called the actual runtime environment values are injected and all configurations are merged into a global runtime manifest.

Sample global runtime manifest produced by the `generate` function:

```json
{
  "remotes": {
    "CraModule/App": {
      "id": "3eb81a0c",
      "url": "http://localhost:3001/remoteEntry.js",
      "scope": "CraModule",
      "module": "./App",
      "metadata": {
        "route": "/cra",
        "displayName": "Example App",
        "displayGroup": "main",
        "order": 10,
        "icon": "ViewList"
      }
    },
    "VanillaModule/Vanilla": {
      "id": "8232ce86",
      "url": "http://localhost:3002/remoteEntry.js",
      "scope": "VanillaModule",
      "module": "./Vanilla",
      "metadata": {
        "route": "/vanilla",
        "displayName": "Example React App",
        "displayGroup": "main",
        "order": 10,
        "icon": "ViewList"
      }
    }
  },
  "modules": {
    "Appshell": {
      "name": "Appshell",
      "shared": {
        "react": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        },
        "react-dom": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        }
      }
    },
    "CraModule": {
      "exposes": {
        "./App": "./src/App"
      },
      "filename": "remoteEntry.js",
      "name": "CraModule",
      "shared": {
        "react": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        },
        "react-dom": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        }
      }
    },
    "VanillaModule": {
      "exposes": {
        "./Vanilla": "./src/App"
      },
      "filename": "remoteEntry.js",
      "name": "VanillaModule",
      "shared": {
        "react": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        },
        "react-dom": {
          "singleton": true,
          "requiredVersion": "^18.2.0"
        }
      }
    }
  }
}
```

This `global runtime manifest` can be consumed by your micro-frontend Appshell host and used to configure the Appshell accordingly.

**What if I want to generate the manifest by a startup script instead?**

> This functionality is exposed by the [@appshell/cli](https://www.npmjs.com/package/@appshell/cli) package. You can simply call `appshell generate manifest --configsDir /path/to/appshell_configs` to produce the global runtime manifest on startup.
