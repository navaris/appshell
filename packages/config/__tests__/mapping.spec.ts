import { entries, values } from 'lodash';
import configmap from '../src/configmap';
import { toAppshellManifest } from '../src/mappers';
import { AppshellConfig, AppshellManifest } from '../src/types';
import config from './assets/appshell.config.json';

describe('mapping configurations to domain objects', () => {
  const APPS_TEST_URL = 'http://remote-endpoint.com';
  const APPS_TEST_REMOTE_ENTRY_PATH = 'remoteEntry.js';

  let appshellManifest: AppshellManifest;
  beforeEach(() => {
    process.env.APPS_TEST_URL = APPS_TEST_URL;
    process.env.APPS_TEST_REMOTE_ENTRY_PATH = APPS_TEST_REMOTE_ENTRY_PATH;
    const configMap = configmap.create(config as AppshellConfig);
    appshellManifest = toAppshellManifest(config as AppshellConfig, configMap);
  });

  it('should match snapshot', () => {
    expect(appshellManifest).toMatchSnapshot();
  });

  it('should replace url placeholders with environment variables', () => {
    expect(
      values(appshellManifest.remotes).some((remote) => remote.url.includes(APPS_TEST_URL)),
    ).toBeTruthy();
  });

  it('should create the scope value from the remote key', () => {
    expect(
      entries(appshellManifest.remotes).every(([key, remote]) => key.includes(remote.scope)),
    ).toBeTruthy();
  });

  it('should create the module value from the remote key', () => {
    expect(
      entries(appshellManifest.remotes).every(([key, remote]) => {
        const moduleName = key.replace(/\/.+/, '');
        const moduleKey = key.replace(moduleName, '.');

        return moduleKey === remote.module;
      }),
    ).toBeTruthy();
  });

  it('should map shareScope to manifest remote', () => {
    expect(values(appshellManifest.remotes).every((remote) => remote.shareScope)).toBeTruthy();
  });
});
