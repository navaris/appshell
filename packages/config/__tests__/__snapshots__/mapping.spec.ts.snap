// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping configurations to domain objects should match snapshot 1`] = `
{
  "modules": {
    "TestModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "remotes": {
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "scope": "BizModule",
      "shareScope": "@sharedscope",
      "url": "http://localhost:4040/remoteEntry.js",
    },
    "TestModule/Bar": {
      "id": "TestModule/Bar",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "module": "./Bar",
      "scope": "TestModule",
      "shareScope": "@sharedscope",
      "url": "http://remote-endpoint.com/remoteEntry.js",
    },
    "TestModule/Foo": {
      "id": "TestModule/Foo",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Foo",
      "scope": "TestModule",
      "shareScope": "@sharedscope",
      "url": "http://remote-endpoint.com",
    },
  },
}
`;
